builtin/logical/database/versioning_large_test.go

```diff
@@ -0,0 +1,317 @@
+				"connection_url":    "sample_connection_url",
+				"plugin_name":       "mock-v5-database-plugin",
+				"verify_connection": true,
+				"allowed_roles":     []string{"*"},
+				"name":              "mockv5",
+				"username":          "mockv5-user",
+				"password":          "mysecurepassword",
+			},
+			assertDynamicUsername: assertStringPrefix("mockv5_user_"),
+			assertDynamicPassword: assertStringRegex("^[a-zA-Z0-9-]{20}"),
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			cleanupReqs := []*logical.Request{}
+			defer cleanup(t, b, cleanupReqs)
+
+			// /////////////////////////////////////////////////////////////////
+			// Configure
+			req := &logical.Request{
+				Operation: logical.CreateOperation,
+				Path:      fmt.Sprintf("config/%s", test.dbName),
+				Storage:   config.StorageView,
+				Data:      test.configData,
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+			defer cancel()
+
+			resp, err := b.HandleRequest(ctx, req)
+			assertErrIsNil(t, err)
+			assertRespHasNoErr(t, resp)
+			assertNoRespData(t, resp)
+
+			cleanupReqs = append(cleanupReqs, &logical.Request{
+				Operation: logical.DeleteOperation,
+				Path:      fmt.Sprintf("config/%s", test.dbName),
+				Storage:   config.StorageView,
+			})
+
+			// /////////////////////////////////////////////////////////////////
+			// Rotate root credentials
+			req = &logical.Request{
+				Operation: logical.UpdateOperation,
+				Path:      fmt.Sprintf("rotate-root/%s", test.dbName),
+				Storage:   config.StorageView,
+			}
+			ctx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
+			defer cancel()
+
+			resp, err = b.HandleRequest(ctx, req)
+			assertErrIsNil(t, err)
+			assertRespHasNoErr(t, resp)
+			assertNoRespData(t, resp)
+
+			// /////////////////////////////////////////////////////////////////
+			// Dynamic credentials
+
+			// Create role
+			dynamicRoleName := "dynamic-role"
+			req = &logical.Request{
+				Operation: logical.UpdateOperation,
+				Path:      fmt.Sprintf("roles/%s", dynamicRoleName),
+				Storage:   config.StorageView,
+				Data: map[string]interface{}{
+					"db_name":     test.dbName,
+					"default_ttl": "5s",
+					"max_ttl":     "1m",
+				},
+			}
+			ctx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
+			defer cancel()
+
+			resp, err = b.HandleRequest(ctx, req)
+			assertErrIsNil(t, err)
+			assertRespHasNoErr(t, resp)
+			assertNoRespData(t, resp)
+
+			cleanupReqs = append(cleanupReqs, &logical.Request{
+				Operation: logical.DeleteOperation,
+				Path:      fmt.Sprintf("roles/%s", dynamicRoleName),
+				Storage:   config.StorageView,
+			})
+
+			// Generate credentials
+			req = &logical.Request{
+				Operation: logical.ReadOperation,
+				Path:      fmt.Sprintf("creds/%s", dynamicRoleName),
+				Storage:   config.StorageView,
+			}
+			ctx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
```

Comment:
Could context cancelation due to timeout mask a request/operation that would have failed otherwise? https://play.golang.org/p/5iqBODFFFoC

Reply:
As far as I know, a timeout could occur prior to a non-timeout error being returned. In this case, the `err` from `HandleRequest` will be non-nil and be caught by `assertErrIsNil`. I can't think of a way that a timeout would occur that we don't get an error back. The example you have requires that the context cancellation (in this case due to the timeout) doesn't return an error which as far as I can tell isn't the behavior of any of the handlers. Ultimately the timeouts here are to help ensure the test doesn't sit there and do nothing for an extended period of time.

Reply:
Context cancelation due to timeout could result from an error within the operation handler's storage calls (if not by something else before that). The failure wouldn't really be telling us that it's a failure of the database implementation itself, but rather something related to storage calls being inaccessible. That's less likely to happen in this case since we're using inmem storage, so it's unlikely that we will hit this timeout even on CI.

Reply:
Agreed. This should really only happen in two scenarios: CI is *really* slow or there's something wrong in the handler(s) that needs to be investigated.